### ISSUE TITLE
Opening .lnk opens destination file for editing

### ISSUE THREAD

**Issue by levicki** [notepad-plus-plus/notepad-plus-plus#9643]
### Description of the Issue
As the title says, opening `.lnk` file using Notepad++ opens shortcut's destination file for editing.

### Steps to Reproduce the Issue
1. Navigate to `C:\ProgramData\Microsoft\Windows\Start Menu\Programs\Administrative Tools` using Explorer
2. Right-click `Registry Editor.lnk` and select `Edit with Notepad++`.

### Expected Behavior
`Registry Editor.lnk` file is opened for editing.

### Actual Behavior
`%windir%\regedit.exe` file is opened for editing instead.

### Debug Information
Notepad++ v7.9.1   (64-bit)
Build time : Nov  2 2020 - 01:07:46
Path : C:\Program Files\Notepad++\notepad++.exe
Admin mode : ON
Local Conf mode : OFF
OS Name : Windows 10 Enterprise (64-bit) 
OS Version : 2009
OS Build : 19042.867
Current ANSI codepage : 1252
Plugins : mimeTools.dll NppConverter.dll nppcrypt.dll NppExport.dll NPPJSONViewer.dll NppMarkdownPanel.dll 

I see that there was an issue #8483 where someone requested this functionality.

If you wonder why would anyone want to edit `.lnk` file itself, I have at least two use cases:

1. To delete its contents (i.e. to make a 0-byte file), and set it to hidden/system/readonly so that rude program installers and updaters cannot litter my start menu with shortcuts I don't need or use.
2. To copy links of special shortcuts that are otherwise not possible to copy because they are grayed in shortcut properties dialog.

I also believe that a text editor should not be opening `.exe` files by default (unless it specifically supports modifications to `PE/PE+` executable structure), because people can accidentally change and save them and thus break their programs or worse yet their Windows install.

If there is no option to turn this off then please consider it as a feature request.

If adding an option is too much of a hassle, then please at least exclude shortcuts to `*.exe` from opening the target instead of shortcut itself.

**Comment by sasumner** [comment]
> a text editor should not be opening .exe files by default

The argument can be made that one really should not open/edit ANY binary file type with a text editor.
This would include .exe files AND .lnk files as both are binary formats.

>make a 0-byte file

There are probably better ways to do this than opening a file and deleting all of its content as if it were text.
An old-timer's example might be `rem.>myfile.lnk` from the venerable command-prompt.

>To copy links of special shortcuts that are otherwise not possible to copy because they are grayed in shortcut properties dialog

I don't have a great answer here, except that clearly portions of a .lnk file are binary, and thus it is arguable whether or not the file should be changed at all by a user using Notepad++.

> exclude shortcuts to *.exe from opening the target instead of shortcut

For one, I'm unconvinced that user needs protection here.
When you open a file, you expect to see certain data (especially if your intention is to edit instead of view).
If you don't see what you expect, you don't just blindly start editing, you seek out the reasons why things appear strange.
In this case, in short order you notice (from the title bar or the tab name) that you have opened an .exe.
Plus, you see that the file is binary, and warning bells against any editing should be going off in your head.

**Comment by levicki** [comment]
> The argument can be made that one really should not open/edit ANY binary file type with a text editor.

I agree.
> This would include .exe files AND .lnk files as both are binary formats.

Yes, but it is far easier even for novice users to restore accidentally corrupted shortcut than to restore random corrupted executable file.

> There are probably better ways to do this than opening a file and deleting all of its content as if it were text.

I am well aware of that, but this way which I find rather convenient used to work before the change which seems to be mostly made to emulate behavior of other broken programs.

> I don't have a great answer here, except that clearly portions of a .lnk file are binary, and thus it is arguable whether or not the file should be changed at all by a user using Notepad++.

To clarify, for that second use case I am only copying text portion to clipboard without modification of the file.

> For one, I'm unconvinced that user needs protection here.

I disagree -- at the minimum, Notepad++ should not follow the shortcut to file types that are not registered for opening using Notepad++.

The users can already override file associations anyway in settings so if they want to open .exe by opening a shortcut to it so be it, but it should be a conscious choice.

Please consider some way of making this new behavior optional.

Obligatory XKCD:

https://xkcd.com/1172/

**Comment by levicki** [comment]
> An old-timer's example might be rem.>myfile.lnk from the venerable command-prompt

That works fine but it has a lot more steps, not to mention that most shortcuts have more than one space in the filename.

This change totally prevents me from opening .lnk files using Notepad++, making the Edit with Notepad++ context menu entry behavior inconsistent.

TL;DR -- .lnk is a file and as such it shouldn't have different treatment. If people want to open arbitrary text files burried somewhere deep on their drives using links, they should create filesystem links using mklink instead of using shortcut files for that and then they will have the desired behavior.

**Comment by levicki** [comment]
> An old-timer's example might be rem.>myfile.lnk from the venerable command-prompt

That works fine in your example, but it involves more steps like opening command prompt in the right folder, not to mention that most shortcut files have more than one space in the filename meaning you have to mess with quotes as well.

This change totally prevents me from opening `.lnk` files using Notepad++, making the **Edit with Notepad++** context menu entry behavior inconsistent.

Also, if Notepad++ supported Linux, would you do the same thing there with desktop shortcut files? The fact that desktop shortcut file is binary under Windows shouldn't make it OK to treat it differently if you already allow opening other binary files.

Besides, `.lnk` is a desktop shortcut **file**, and as such it shouldn't have a different treatment from the editor unless that behavior is optional.

If people want to open arbitrary text files burried somewhere deep on their drives using links, then they can simply create **filesystem links** using OS provided `mklink` command (or using free [Link Shell Extension](https://schinagl.priv.at/nt/hardlinkshellext/linkshellextension.html) context menu addon if they prefer GUI) instead of using desktop shortcut files. Unlike `.lnk` files, filesystem links do not require adding any special handling code to the editor.

That way we all get the desired behavior -- they get links that directly open destination file when opened in Notepad++, and weirdos like me can still edit shortcut files.

Please reconsider reverting the change or at least adding an option to turn it off.

**Comment by ccuser44** [comment]
Yeah this might be sometimes problematic.

**Comment by sasumner** [comment]
Sorry, but it isn't going to be reconsidered.

**Comment by levicki** [comment]
> Sorry, but it isn't going to be reconsidered.

I heard that's more or less the same answer China gave when it comes to Hong Kong freedom.

If only that was solvable with a simple checkbox in Settings like this issue you won't reconsider...

**Comment by Vizepi** [comment]
@sasumner Have you considered that notepad++ can be extended by plugins and that manipulating .lnk files is something that such plugins can enable? You're clearly being lazy in your decision.
